update instructions for building operator-sdk

elmiko opened the pull request:
<!--

Before making a PR, please read our contributing guidelines https://github.com/operator-framework/operator-sdk/blob/master/CONTRIBUTING.MD

Note: Make sure your branch is rebased to the latest upstream master.

-->

**Description of the change:**

this change updates the minimum required go lang version in one of the developer docs, and also changes the method for downloading the source code. it should produce fewer warnings for developers wishing to install operator-sdk from source. it also adds a note about needing GOPROXY set to its default.


**Motivation for the change:**

Inspired by warning messages found while investigating issue #2573, the proposed changes here produce fewer warnings during installation for developers.

<!--

Note: If this PR is fixing an issue make sure to add a note saying:
Closes #<ISSUE_NUMBER>

-->

camilamacedo86 commented on doc/user/install-operator-sdk.md:
````diff
@@ -86,8 +86,8 @@ $ chmod +x operator-sdk-${RELEASE_VERSION}-x86_64-apple-darwin && sudo mkdir -p
 ```sh
-$ go get -d github.com/operator-framework/operator-sdk # This will download the git repository and not install it
-$ cd $GOPATH/src/github.com/operator-framework/operator-sdk
+$ git clone https://github.com/operator-framework/operator-sdk
````
Users do not need to do the git clone. This would be the recommendation for who would like to contribute with and not just use.

camilamacedo86 replied:
```suggestion
$ git clone https://github.com/operator-framework/operator-sdk $GOPATH/src/github.com/operator-framework/operator-sdk
```

camilamacedo86 commented on doc/dev/developer_guide.md:
```diff
@@ -4,7 +4,7 @@ This document explains how to setup your dev environment.
 ## Prerequisites
 - [git][git-tool]
-- [go][go-tool] version v1.12+
+- [go][go-tool] version v1.13+
```
👍

camilamacedo86 commented on doc/user/install-operator-sdk.md:
```diff
@@ -86,8 +86,8 @@ $ chmod +x operator-sdk-${RELEASE_VERSION}-x86_64-apple-darwin && sudo mkdir -p
-$ go get -d github.com/operator-framework/operator-sdk # This will download the git repository and not install it
-$ cd $GOPATH/src/github.com/operator-framework/operator-sdk
+$ git clone https://github.com/operator-framework/operator-sdk
+$ cd operator-sdk
```
```suggestion
$ cd $GOPATH/src/github.com/operator-framework/operator-sdk
```

elmiko replied:
hmm, building from within the `$GOPATH` is what seems to be causing problems for me though. i wonder if we need another step here about the proper way to configure the go modules?

when i look at `go help modules`, the advice there is to clone a repo with modules outside of the `$GOPATH` tree. it seemed like this would be the easiest way for someone to either build the tool for usage or to contribute. i will dig a little deeper into the go docs to see if there is an option to relax the go module awareness.

camilamacedo86 replied:
It is solved .. closing as resoolved

elmiko commented:
@camilamacedo86 ok, updated the install instructions. i think this is actually unrelated to the errors i was seeing from #2573, but at the least the `git clone` will reduce the warnings produced by `go get`. thanks again for the reviews =)

elmiko commented on doc/user/install-operator-sdk.md:
````diff
@@ -86,7 +86,7 @@ $ chmod +x operator-sdk-${RELEASE_VERSION}-x86_64-apple-darwin && sudo mkdir -p
 
 ```sh
-$ go get -d github.com/operator-framework/operator-sdk # This will download the git repository and not install it
+$ git clone https://github.com/operator-framework/operator-sdk $GOPATH/src/github.com/operator-framework/operator-sdk
````
ok, this is what i had originally. i can revert those lines.

elmiko commented:
@estroz @camilamacedo86 i changed the instructions to something that takes the modules approach into account, also makes it easier for new devs. i also added a note about some linux distributions needing the `GOPROXY` to be set as this will cause issues for some users.

estroz commented on doc/user/install-operator-sdk.md:
```diff
@@ -86,13 +86,17 @@ $ chmod +x operator-sdk-${RELEASE_VERSION}-x86_64-apple-darwin && sudo mkdir -p
 
+Please note, on some Linux distributions you will need to set your `GOPROXY`
+environment variable. Ensure that it is set to the following,
+`GOPROXY="https://proxy.golang.org,direct"`.
```
I don't think this is a distribution-specific requirement. Why do you think it is out of curiosity?

elmiko replied:
based on my experience with the default go installation in fedora, and these two comments on the issue i raised:

https://github.com/operator-framework/operator-sdk/issues/2573#issuecomment-587089134
> In the GO 1.13+, the GOPROXY is by default `https://proxy.golang.org,direct` so because of this has no need to add in the docs. For someone reason, your go env was not set up properly.

https://github.com/operator-framework/operator-sdk/issues/2573#issuecomment-587091615
> on Fedora 31, you are required to have GOPROXY set, this is due to the way golang is packaged on Fedora, I found this out when doing the 1.15.0 release on my Fedora server.

it leads me to believe that on some distributions (eg the one that @camilamacedo86 is using), the `GOPROXY` is set by default to the location referenced in this pr. it certainly does not seem to be set that way on fedora 31, curiously enough i think this might work on earlier versions of fedora.

i can make that statement more generic though, just advising to check your `GOPROXY`, seems like sound advice regardless. i spent several hours scratching my head and looking for the root cause on issue #2574 due to fedora's configuration.

camilamacedo86 replied:
WDYT about ?

```suggestion
**Note:**  Ensure that your `GOPROXY` is set with its default values for go versions 1.13+ which is `https://proxy.golang.org,direct`
```

@elmiko, 

Note that I know that the GOPROXY has these values by default for go 1.13+ and I do NOT know why you did not have it in your env but also I am afraid that it has NO relation with the distro.

elmiko replied:
++, i'm fine with that change.

my only mention of the distro is that it looks like fedora explicitly _does not_ set GOPROXY. my thought was that some distros (perhaps the one you use) do set it. regardless, i agree that the docs should be more generic in nature with regards to this issue.

estroz replied:
This might be for certain package managers or packaging repositories, not distro-specific. Regardless I agree with this change.
